File: ruby/acronym/acronym.rb
# Acronym generator
class Acronym
  VERSION = 1

  def self.abbreviate(phrase)
    phrase.split(/\W+/).flat_map { |word| split_camelcase(word) }.join.upcase
  end

  def self.split_camelcase(word)
    if word =~ /^[[:upper:]]+$/
      word[0]
    else
      word.split(/(?=[A-Z])/).map { |w| w[0] }
    end
  end
end

File: elixir/acronym/acronym.exs
defmodule Acronym do
  @doc """
  Generate an acronym from a string. 
  "This is a string" => "TIAS"
  """
  @spec abbreviate(string) :: String.t()
  def abbreviate(string) do
    string
    |> String.split(~r{[^\w]}, trim: true)
    |> Enum.flat_map(&split_camel_case/1)
    |> Enum.map(&String.first/1)
    |> Enum.join
    |> String.upcase
  end

  defp split_camel_case(string) do
    string
    |> String.split(~r{(?<!^)(?=[A-Z])}, trim: true)
  end
end
